gnat2goto/driver/tree_walk.adb

```diff
@@ -1007,109 +1008,254 @@ package body Tree_Walk is
    -- Do_Array_Assignment --
    -------------------------
 
-   --  TODO: multi-dimensional arrays
-   function Do_Array_Assignment (LHS_Expr : Irep; RHS_Expr : Irep; N : Node_Id)
-                                return Irep
+   --  The following function builds a generalised array assignment
+   --  dest := src_1 & src_2 & .. & src_n
+   --  for $n>0$, src_i may overlap with dest
+   --  Let ArrT := struct { int first; int last; T* data; }
+   ----------------------------------------------------------------------------
+   --  void concat_assign(ArrT dest, ArrT src_1, ArrT src_2, .., ArrT src_n) {
+   --    offset_step = 0;
+   --    slice_size = src_1.last - src_1.first + 1;
+   --    slice_temp = (T*)malloc(slice_size * sizeof(T));
+   --    memcpy(slice_temp, src_1.data, slice_size * sizeof(T));
+   --    memcpy(dest.data + offset_step, slice_temp, slice_size * sizeof(T));
```

Comment:
I'd suggest clarifying that the temporary is required due the scenario of an array copying a slice back to itself

Reply:
Added.